tools/shared/src/test/scala/org/scalajs/core/tools/linker/AnalyzerTest.scala

```diff
@@ -0,0 +1,240 @@
+package org.scalajs.core.tools.linker
+
+import org.junit.Test
+import org.junit.Assert._
+
+import org.scalajs.core.ir
+import org.scalajs.core.ir.ClassKind
+import org.scalajs.core.ir.Definitions._
+import org.scalajs.core.ir.Trees._
+import org.scalajs.core.ir.Types._
+
+import org.scalajs.core.tools.logging.NullLogger
+import org.scalajs.core.tools.io._
+import org.scalajs.core.tools.linker._
+import org.scalajs.core.tools.linker.analyzer._
+import org.scalajs.core.tools.linker.standard._
+
+import Analysis._
+
+import org.scalajs.core.tools.linker.testutils._
+
+class AnalyzerTest {
+  private val reqsFactory = SymbolRequirement.factory("unit test")
+
+  private val emptyOptHints: OptimizerHints = OptimizerHints.empty
+
+  implicit private val noPosition: ir.Position = ir.Position.NoPosition
+
+  private val fromAnalyzer = FromCore("analyzer")
+  private val fromUnitTest = FromCore("unit test")
+
+  private def trivialCtor(enclosingClassName: String) = {
+    MethodDef(static = false, Ident("init___"), Nil, NoType,
+        Some(ApplyStatically(This()(ClassType(enclosingClassName)),
+            ClassType(ObjectClass), Ident("init___"), Nil)(NoType)))(
+        emptyOptHints, None)
+  }
+
+  @Test
+  def trivialOK(): Unit = {
+    val analysis = computeAnalysis(Nil)
+    assertNoError(analysis)
+  }
+
+  @Test
+  def missingJavaLangObject(): Unit = {
+    val analysis = computeAnalysis(Nil, stdlib = false)
+    assertHasErrors(analysis,
+        MissingJavaLangObjectClass(fromAnalyzer))
+  }
+
+  @Test
+  def invalidJavaLangObject(): Unit = {
+    val invalidJLObjectDefs = Seq(
+      ClassDef(Ident(ObjectClass), ClassKind.Class, Some(Ident("Lparent")),
+          Nil, None, Nil, Nil)(emptyOptHints),
+      ClassDef(Ident(ObjectClass), ClassKind.ModuleClass, None, Nil, None,
+          Nil, Nil)(emptyOptHints),
+      ClassDef(Ident(ObjectClass), ClassKind.Class, None,
+          List(Ident("Lparent")), None, Nil, Nil)(emptyOptHints)
+    )
+
+    for (jlObjectDef <- invalidJLObjectDefs) {
+      val analysis = computeAnalysis(Seq(jlObjectDef), stdlib = false)
+      assertHasErrors(analysis,
+          InvalidJavaLangObjectClass(fromAnalyzer))
+    }
+  }
+
+  @Test
+  def cycleInInheritanceChainThroughParentClasses(): Unit = {
+    val classDefs = Seq(
+        ClassDef(Ident("LA"), ClassKind.Class, Some(Ident("LB")), Nil, None,
+            Nil, Nil)(emptyOptHints),
+        ClassDef(Ident("LB"), ClassKind.Class, Some(Ident("LA")), Nil, None,
+            Nil, Nil)(emptyOptHints)
+    )
+
+    val analysis = computeAnalysis(classDefs, reqsFactory.classData("LA"))
+
+    assertHasError("CycleInInheritanceChain(LA, LB)", analysis) {
+      case CycleInInheritanceChain(List(a, b), `fromUnitTest`) =>
+        a.encodedName == "LA" && b.encodedName == "LB"
```

Comment:
Double checking: We are OK with imposing the starting point of the inheritance chain in the test here, since we require A to be present and not B (also below).

Also: Consider creating a matcher for `encodedName`.

Reply:
> Double checking

Yes, it's deterministic: the class that is required "from the outside" always appears first in the cycle.